src/GraphQL/Internal/Syntax/AST.hs

```diff
@@ -104,6 +105,12 @@ instance IsString Name where
 instance Aeson.ToJSON Name where
   toJSON = Aeson.toJSON . unName
 
+instance Aeson.FromJSON Name where
+  parseJSON = Aeson.withText "Name" $ \v ->
+    case makeName v of
+      Left _ -> mempty
```

Comment:
Is `mempty` the right thing here? Should we maybe `fail` instead? I'm not even sure why `mempty` is a valid value for `Name` - maybe @jml remembers :)

Reply:
In both this and the other case, you're [running inside the Aeson `Parser` monad](https://hackage.haskell.org/package/aeson-1.2.3.0/docs/Data-Aeson.html#t:FromJSON).  Protolude is a little more aggressive about moving `fail` out of the base `Monad` class, so it looks like I need to `import Control.Monad.Fail` and then it works.

Reply:
@teh This is a wart in Aeson IMO. Because `mempty` wasn't prefixed with `return` (or `pure`), it's `mempty :: Aeson.Parser Name`, not `mempty :: Name` (which wouldn't compile).

If you look at the source for `Parser a`, it has [an implementation of `Monoid`](http://hackage.haskell.org/package/aeson-1.2.3.0/docs/src/Data-Aeson-Types-Internal.html#line-323) such that `mempty = fail "mempty"`. I think this kind of sucks.

`empty` (i.e. failed `Alternative`, not monoid identity), would be much more correct.

But `fail` is best, because it's informative.